cmd/kubeadm/app/phases/upgrade/health.go

```diff
@@ -59,49 +66,150 @@ func (c *healthCheck) Name() string {
 // - all control-plane Nodes are Ready
 // - (if self-hosted) that there are DaemonSets with at least one Pod for all control plane components
 // - (if static pod-hosted) that all required Static Pod manifests exist on disk
-func CheckClusterHealth(client clientset.Interface, ignoreChecksErrors sets.String) error {
-	fmt.Println("[upgrade] Making sure the cluster is healthy:")
+func CheckClusterHealth(client clientset.Interface, cfg *kubeadmapi.ClusterConfiguration, ignoreChecksErrors sets.String) error {
+	fmt.Println("[upgrade] Running cluster health checks")
 
 	healthChecks := []preflight.Checker{
 		&healthCheck{
-			name:   "APIServerHealth",
+			name:   "CreateJob",
 			client: client,
-			f:      apiServerHealthy,
+			cfg:    cfg,
+			f:      createJob,
 		},
 		&healthCheck{
 			name:   "ControlPlaneNodesReady",
 			client: client,
 			f:      controlPlaneNodesReady,
 		},
-		// TODO: Add a check for ComponentStatuses here?
+		&healthCheck{
+			name:   "StaticPodManifest",
+			client: client,
+			cfg:    cfg,
+			f:      staticPodManifestHealth,
+		},
 	}
 
-	healthChecks = append(healthChecks, &healthCheck{
-		name:   "StaticPodManifest",
-		client: client,
-		f:      staticPodManifestHealth,
-	})
-
 	return preflight.RunChecks(healthChecks, os.Stderr, ignoreChecksErrors)
 }
 
-// apiServerHealthy checks whether the API server's /healthz endpoint is healthy
-func apiServerHealthy(client clientset.Interface) error {
-	healthStatus := 0
+// CreateJob is a check that verifies that a Job can be created in the cluster
+func createJob(client clientset.Interface, cfg *kubeadmapi.ClusterConfiguration) (lastError error) {
+	const (
+		jobName = "upgrade-health-check"
+		ns      = metav1.NamespaceSystem
+		timeout = 15 * time.Second
+	)
 
-	// If client.Discovery().RESTClient() is nil, the fake client is used, and that means we are dry-running. Just proceed
+	// If client.Discovery().RESTClient() is nil, the fake client is used.
+	// Return early because the kubeadm dryrun dynamic client only handles the core/v1 GroupVersion.
```

Comment:
had to add this.

the dynamic client fails when trying to get batch/v1 Job here:
https://github.com/kubernetes/kubernetes/blob/fa1ca34550b70b017234099a5d469fd99e7eed0d/cmd/kubeadm/app/util/apiclient/clientbacked_dryrun.go#L77

i think we should stop using the dynamic client for dryrun, but this is a much larger refactor.
until then we have to skip the logic in this preflight check on dryrun.

Reply:
+1 we can have a backlog issue for this so an eager contributor can try it.